File: gdp-docker/gdp/lang/python/gdp/GDP_DATUM.py
#!/usr/bin/env python

# ----- BEGIN LICENSE BLOCK -----
#	GDP: Global Data Plane
#	From the Ubiquitous Swarm Lab, 490 Cory Hall, U.C. Berkeley.
#
#	Copyright (c) 2015-2019, Regents of the University of California.
#	All rights reserved.
#
#	Permission is hereby granted, without written agreement and without
#	license or royalty fees, to use, copy, modify, and distribute this
#	software and its documentation for any purpose, provided that the above
#	copyright notice and the following two paragraphs appear in all copies
#	of this software.
#
#	IN NO EVENT SHALL REGENTS BE LIABLE TO ANY PARTY FOR DIRECT, INDIRECT,
#	SPECIAL, INCIDENTAL, OR CONSEQUENTIAL DAMAGES, INCLUDING LOST
#	PROFITS, ARISING OUT OF THE USE OF THIS SOFTWARE AND ITS DOCUMENTATION,
#	EVEN IF REGENTS HAS BEEN ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
#
#	REGENTS SPECIFICALLY DISCLAIMS ANY WARRANTIES, INCLUDING, BUT NOT
#	LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
#	FOR A PARTICULAR PURPOSE. THE SOFTWARE AND ACCOMPANYING DOCUMENTATION,
#	IF ANY, PROVIDED HEREUNDER IS PROVIDED "AS IS". REGENTS HAS NO
#	OBLIGATION TO PROVIDE MAINTENANCE, SUPPORT, UPDATES, ENHANCEMENTS,
#	OR MODIFICATIONS.
# ----- END LICENSE BLOCK -----


from MISC import *
from GDP_BUF import GDP_BUF

# Formatting options
# GDP_DATUM_PRTEXT        0x00000001      // print data as text
# GDP_DATUM_PRDEBUG       0x00000002      // print debugging info
# GDP_DATUM_PRSIG         0x00000004      // print the signature
# GDP_DATUM_PRQUIET       0x00000008      // don't print any metadata
# GDP_DATUM_PRMETAONLY    0x00000010      // only print metadata
(GDP_DATUM_PRTEXT, GDP_DATUM_PRDEBUG, GDP_DATUM_PRSIG,
        GDP_DATUM_PRQUIET, GDP_DATUM_PRMETAONLY) = (1, 2, 3, 8, 16)


class GDP_DATUM:

    """
    A class only for internal use. The C datum equivalent exposed to
    python users is of a dictionary.

    This is to avoid any side effects that the underlying buffer
    implementation may cause.
    """

    class gdp_datum_t(Structure):
        pass

    def __init__(self, **kwargs):
        """
        Constructor: Two ways of creating a new pythong GDP datum object:
        - Create a new object and allocate new memeory for it by calling
          the C library function
        - Create a new object, but associate it with an existing memory
          location.
        In the first case: we need to free the memory when it is no longer
        needed. In the second case: we shouldn't free memory allocated by
        someone else.
        """
        if len(kwargs) == 0:

            # We do need to create a corresponding C data structure.
            __func = gdp.gdp_datum_new
            __func.argtypes = []
            __func.restype = POINTER(self.gdp_datum_t)

            self.ptr = __func()
            self.did_i_create_it = True

        else:
            # we probably got passed a C pointer to an existing datum.
            if "ptr" in kwargs:
                self.ptr = kwargs["ptr"]
                self.did_i_create_it = False
            else:
                raise Exception     # FIXME

    def __del__(self):
        """
        Destructor: Does nothing if this GDP datum was created by
        passing an exisiting datum pointer
        """

        if self.did_i_create_it:
            __func = gdp.gdp_datum_free
            __func.argtypes = [POINTER(self.gdp_datum_t)]
            __func(self.ptr)


    def print_(self, fh, flags):
        """
        Print the GDP datum C memory location contents to a file handle fh.
        fh could be sys.stdout, or any other open file handle. Flags are
        predefined constants (see GDP_DATUM_PR*)

        Note: This just calls the corresponding C library function which
        handles the printing
        """

        # need to convert this file handle to a C FILE*
        __fh = PyFile_AsFile(fh)
        __func = gdp.gdp_datum_print
        __func.argtypes = [POINTER(self.gdp_datum_t), FILE_P, c_uint32]
        # ignore the return value
        __func(self.ptr, __fh, c_uint32(flags))

    def __getitem__(self, key):

        if key == "recno":
            return self.getrecno()
        if key == "ts":
            return self.getts()
        if key == "dlen":
            return self.getdlen()
        if key == "buf":
            return self.getbuf()
        if key == "sig":
            return self.getsig()
        raise NotImplementedError

    def reset(self):
        """ Reset the datum """
        __func = gdp.gdp_datum_reset
        __func.argtypes = [POINTER(self.gdp_datum_t)]
        __func(self.ptr)


    def gethash(self, gin):
        """ Compute the hash of the datum for a given log? """

        assert isinstance(gin, GDP_GIN)
        __func = gdp.gdp_datum_hash
        __func.argtypes = [POINTER(self.gdp_datum_t),
                                    POINTER(GDP_GIN.gdp_gin_t)]
        __func.restype = POINTER(gdp_hash_t)

        gdp_hash_ptr = __func(self.ptr, gin.ptr)
        return GDP_HASH(ptr=gdp_hash_ptr)


    def getrecno(self):
        """
        Get the corresponding record number associated with this datum
        """

        __func = gdp.gdp_datum_getrecno
        __func.argtypes = [POINTER(self.gdp_datum_t)]
        __func.restype = gdp_recno_t

        ret = __func(self.ptr)
        return int(ret)

    def getts(self):
        """
        Return the timestamp associated with this GDP_DATUM in the form of a
        dictionary. The keys are: tv_sec, tv_nsec and tv_accuracy
        """

        ts = EP_TIME_SPEC()
        __func = gdp.gdp_datum_getts
        __func.argtypes = [
            POINTER(self.gdp_datum_t), POINTER(EP_TIME_SPEC)]
        # ignore the return value

        __func(self.ptr, byref(ts))
        # represent the time spec as a dictionary

        ret = {}
        ret['tv_sec'] = ts.tv_sec
        ret['tv_nsec'] = ts.tv_nsec
        ret['tv_accuracy'] = ts.tv_accuracy

        return ret

    def getdlen(self):
        """
        Returns the length of the data in GDP_BUF associated with
        this GDP_DATUM.
        """

        __func = gdp.gdp_datum_getdlen
        __func.argtypes = [POINTER(self.gdp_datum_t)]
        __func.restype = c_size_t

        ret = __func(self.ptr)
        return ret

    def getbuf(self):
        """ Return the GDP_BUF associated with this GDP_DATUM. """

        __func = gdp.gdp_datum_getbuf
        __func.argtypes = [POINTER(self.gdp_datum_t)]
        __func.restype = POINTER(GDP_BUF.gdp_buf_t)

        gdp_buf_ptr = __func(self.ptr)
        gdp_buf = GDP_BUF(ptr=gdp_buf_ptr)
        gdp_buf.datum = self    # garbage collection workaround
        return gdp_buf


    # def setbuf(self, data):
    #     "Set the buffer to the given data. data is a python string"

    #     __func = gdp.gdp_datum_getbuf
    #     __func.argtypes = [POINTER(self.gdp_datum_t)]
    #     __func.restype = POINTER(GDP_BUF.gdp_buf_t)

    #     gdp_buf_ptr = __func(self.ptr)

    #     __func_write = gdp.gdp_buf_write
    #     __func_write.argtypes = [POINTER(GDP_BUF.gdp_buf_t), c_void_p, c_size_t]
    #     __func_write.restype = c_int

    #     size = c_size_t(len(data))
    #     tmp_buf = create_string_buffer(data, len(data))
    #     written_bytes = __func_write(gdp_buf_ptr, byref(tmp_buf), size)
    #     return


    def getsig(self):
        """ Return the signature as a binary string.
        As of current, the signature is over (recno|data)"""

        # get a pointer to signature buffer
        __func = gdp.gdp_datum_getsig
        __func.argtypes = [POINTER(self.gdp_datum_t)]
        __func.restype = POINTER(GDP_BUF.gdp_buf_t)

        sig_buf_ptr = __func(self.ptr)

        return GDP_SIG(ptr=sig_buf_ptr)

File: combined-docker/gdp/lang/java/org/terraswarm/gdp/GDP_DATUM.java
/* Global Data Plane Utility Datum

   Copyright (c) 2015-2016 The Regents of the University of California.
   All rights reserved.
   Permission is hereby granted, without written agreement and without
   license or royalty fees, to use, copy, modify, and distribute this
   software and its documentation for any purpose, provided that the above
   copyright notice and the following two paragraphs appear in all copies
   of this software.

   IN NO EVENT SHALL THE UNIVERSITY OF CALIFORNIA BE LIABLE TO ANY PARTY
   FOR DIRECT, INDIRECT, SPECIAL, INCIDENTAL, OR CONSEQUENTIAL DAMAGES
   ARISING OUT OF THE USE OF THIS SOFTWARE AND ITS DOCUMENTATION, EVEN IF
   THE UNIVERSITY OF CALIFORNIA HAS BEEN ADVISED OF THE POSSIBILITY OF
   SUCH DAMAGE.

   THE UNIVERSITY OF CALIFORNIA SPECIFICALLY DISCLAIMS ANY WARRANTIES,
   INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
   MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE. THE SOFTWARE
   PROVIDED HEREUNDER IS ON AN "AS IS" BASIS, AND THE UNIVERSITY OF
   CALIFORNIA HAS NO OBLIGATION TO PROVIDE MAINTENANCE, SUPPORT, UPDATES,
   ENHANCEMENTS, OR MODIFICATIONS.

   PT_COPYRIGHT_VERSION_2
   COPYRIGHTENDKEY

 */

package org.terraswarm.gdp; 

import java.nio.ByteBuffer;
import java.nio.IntBuffer;
import java.util.Date;
import java.util.Map;
import java.awt.PointerInfo;
import java.lang.Exception;


import com.sun.jna.Native;
import com.sun.jna.Memory;
import com.sun.jna.Pointer;
import com.sun.jna.ptr.PointerByReference;
import org.terraswarm.gdp.NativeSize; // Fixed by cxh in makefile.

/**
 * A GDP Datum. The original C GDP Datum may create confusion since
 * data is stored in a buffer (which unsurprisingly acts like a
 * buffer).
 *
 * @author nitesh mor
 *
 */
public class GDP_DATUM {

    /**
     * Create a new datum.
     */
    public GDP_DATUM() {
        this.gdp_datum_ptr = Gdp21Library.INSTANCE.gdp_datum_new();
        this.did_i_create_it = true;
        return;
    }
    
    /**
     * Use an already allocated C data structure.
     * @param p Pointer to existing C memory of type gdp_datum_t
     */    
    public GDP_DATUM(PointerByReference d) {
        this.gdp_datum_ptr = d;
        this.did_i_create_it = false;
        return;
    }
    
    /**
     * If we allocated memory ourselves, free it.
     */
    public void finalize() { 
        if (this.did_i_create_it == true) {
            Gdp21Library.INSTANCE.gdp_datum_free(this.gdp_datum_ptr);
        }
    }
    

    public void reset() {
        Gdp21Library.INSTANCE.gdp_datum_reset(this.gdp_datum_ptr);
    }

    // TODO implement gdp_datum_copy

    // TODO implement gdp_datum_hash

    // TODO implement gdp_datum_hash_equal

    public void vrfy(GDP_GIN gin) throws GDPException{
        EP_STAT estat;
        estat = Gdp21Library.INSTANCE.gdp_datum_vrfy(
                        this.gdp_datum_ptr, gin.ginh);
        GDP.check_EP_STAT(estat, "gdp_datum_vrfy failed");
    }

    /**
     * Get the associated record number.
     * @return Record number
     */
    public long getrecno() {
        long recno = Gdp21Library.INSTANCE.gdp_datum_getrecno(
                                        this.gdp_datum_ptr);
        return recno;
    }
    
    /**
     * Get the timestamp.
     * @return timestamp associated with the C data structure
     */
    public EP_TIME_SPEC getts() {
        EP_TIME_SPEC ts = new EP_TIME_SPEC();
        Gdp21Library.INSTANCE.gdp_datum_getts(this.gdp_datum_ptr, ts);
        return ts;
    }
    
    /**
     * Get the length of data in the bufffer.
     * @return Length of buffer
     */
    public NativeSize getdlen() {
        PointerByReference buf = Gdp21Library.INSTANCE.gdp_datum_getbuf(
                                        this.gdp_datum_ptr);
        NativeSize len = Gdp21Library.INSTANCE.gdp_buf_getlength(buf);
        return len;
    }
    
    /**
     * Get the data in the buffer. Internally, queries the buffer and
     * effectively drains it as well.
     * @return the bytes in the buffer or null if the pointer to the
     * buffer is null.
     */
    public byte[] getbuf(){
        PointerByReference buf = Gdp21Library.INSTANCE.gdp_datum_getbuf(
                this.gdp_datum_ptr);
        NativeSize len = Gdp21Library.INSTANCE.gdp_buf_getlength(buf);
        Pointer bufptr = Gdp21Library.INSTANCE.gdp_buf_getptr(buf, len);
        if (bufptr == null) {
            return null;
        }
        byte[] bytes = bufptr.getByteArray(0, len.intValue());
        return bytes;
    }
    
    
    /**
     * Set the buffer to specified bytes.
     * @param data The data
     */
    public void setbuf(byte[] data){
        
        // Create a C string that is a copy (?) of the Java String.
        Memory memory = new Memory(data.length);
        // FIXME: not sure about alignment.
        Memory alignedMemory = memory.align(4);
        memory.clear();
        Pointer pointer = alignedMemory.share(0);
        
        for (int i=0; i<data.length; i++) {
            pointer.setByte(i,data[i]);
        }

        // Now feed this data into the gdp buffer
        PointerByReference dbuf = Gdp21Library.INSTANCE.gdp_datum_getbuf(
                                    this.gdp_datum_ptr);
        Gdp21Library.INSTANCE.gdp_buf_write(dbuf, pointer, 
                                new NativeSize(data.length));
    }

    // TODO implement gdp_datum_getsig

    // The associated C data structure.
    public PointerByReference gdp_datum_ptr = null;

    // To keep track of whether we need to free the associated C memory.
    private boolean did_i_create_it = false;
    

}
